iothub_client/src/iothubtransporthttp.c

```diff
@@ -2262,57 +2234,59 @@ static void DoMessages(HTTPTRANSPORT_HANDLE_DATA* handleData, HTTPTRANSPORT_PERD
-                                                LogError("failed to assemble callback info");
-                                                if (!abandonOrAcceptMessage(handleData, deviceData, etagValue, IOTHUBMESSAGE_ABANDONED))
-                                                {
-                                                    LogError("HTTP Transport layer failed to report ABANDON disposition");
-                                                }
+                                                LogError("failed to create disposition context");
+                                                abandon = true;
                                             }
                                             else
                                             {
-                                                bool abandon;
-                                                if (handleData->transport_callbacks.msg_cb(messageData, deviceData->device_transport_ctx))
+                                                if (IoTHubMessage_SetDispositionContext(receivedMessage, dispositionContext, DestroyMessageDispositionContext) != IOTHUB_MESSAGE_OK)
                                                 {
-                                                    abandon = false;
+                                                    LogError("Failed setting disposition context in IOTHUB_MESSAGE_HANDLE");
+                                                    DestroyMessageDispositionContext(dispositionContext);
+                                                    abandon = true;
                                                 }
-                                                else
+                                                else if (!handleData->transport_callbacks.msg_cb(receivedMessage, deviceData->device_transport_ctx))
                                                 {
                                                     LogError("IoTHubClientCore_LL_MessageCallback failed");
                                                     abandon = true;
                                                 }
-
-                                                /*Codes_SRS_TRANSPORTMULTITHTTP_17_096: [If IoTHubClientCore_LL_MessageCallback returns false then _DoWork shall "abandon" the message.] */
-                                                if (abandon)
-                                                {
-                                                    (void)IoTHubTransportHttp_SendMessageDisposition(messageData, IOTHUBMESSAGE_ABANDONED);
-                                                }
                                             }
                                         }
-                                        IoTHubMessage_Destroy(receivedMessage);
+
+                                        if (abandon)
+                                        {
+                                            IoTHubMessage_Destroy(receivedMessage);
```

Comment:
The same way you added MQTT comment comments about ownership, recommend something along lines of
```c
// If IoTHubMessage_SetDispositionContext succeeds above, it transitions ownership of dispositionContext to the receivedMessage.  IoTHubMessage_Destroy() below will handle freeing it.
```

Reply:
Done.